bpf/process/bpf_fork.c

```diff
@@ -44,7 +44,6 @@ BPF_KPROBE(event_wake_up_new_task, struct task_struct *task)
 		curr->key.pid = tgid;
 		curr->key.ktime = ktime_get_ns();
 		curr->nspid = get_task_pid_vnr();
-		curr->binary = parent->binary;
```

Comment:
Is this okay @tpapagian? Naively, I replaced that line with `curr->bin = parent->bin` and it broke my tests, I'm not exactly sure why. I guess maybe we should copy the complete struct with the binary path inside of it?

Reply:
I believe that we need to setup `bin` here (maybe with a `memcpy`?). An example that I am thinking of is having a program that does a `fork` but not an `exec`. 

I have tried a program like [this](https://gist.github.com/tpapagian/19d708e9319020854d76289809928bc0) that does a `fork` (and not an `exec`) and the child process reads `/etc/passwd`. I used [this](https://gist.github.com/tpapagian/70f0b15665b00e446d0a3f37b32ddf98) tracing policy. I expect that adding a `matchBinaries` selector in that tracing policy not to work because the child will not have the proper binary name. 

My understanding is that `bin` is used only for `matchBinaries`, as without that I got a correct event similar to [this](https://gist.github.com/tpapagian/f1abf29cca711b38dce846cedf03c9b6).

I will do a full review to understand the whole approach. Sorry if I miss anything.

Reply:
For some obscure reasons, this breaks the feature and I can't understand (yet) why:

```diff
diff --git a/bpf/process/bpf_fork.c b/bpf/process/bpf_fork.c
index 2d36be4f..55ffe2d7 100644
--- a/bpf/process/bpf_fork.c
+++ b/bpf/process/bpf_fork.c
@@ -58,6 +58,7 @@ BPF_KPROBE(event_wake_up_new_task, struct task_struct *task)
        curr->key.pid = tgid;
        curr->key.ktime = ktime_get_ns();
        curr->nspid = get_task_pid_vnr();
+       memcpy(&curr->bin, &parent->bin, sizeof(curr->bin));
        curr->pkey = parent->key;
 
        /* Setup the msg_clone_event and sent to the user. */
```

Reply:
Yes, initially this sounds a bit strange to me as well. After looking a bit more into that, it seems that you need to set the binary path to 0 before doing `curr->bin.path_length = probe_read_str(curr->bin.path, BINARY_PATH_MAX_LEN, &p->args)` in `execve_send`. 

Without the `memcpy` in clone event `curr->bin.path` is always zeroed. With the `memcpy`, you set an initial value to that buffer during the clone event. The `probe_read_str` call in `execve_send` only writes the first `N` bytes of the buffer where `N` is the length of the binary. If the binary length during `clone` is larger than the binary length during `exec` we still have some random bytes at the end of the buffer. In that case, we will never match in the map during `matchBinaries`. This seems to be the case in the tests. 

I believe a simple fix is to call something like `memset(&(curr->bin), 0, sizeof(struct binary))` before `curr->bin.path_length = probe_read_str(curr->bin.path, BINARY_PATH_MAX_LEN, &p->args)`.